### PR TITLE
CC-related improvements

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**

In this PR, I've updated some CC-related types, and I've implemented a first try for CC DIIS. I'm still experimenting with the implementation, but I'd like to merge some work to develop already, so I can rebuild production containers.

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files

### CODE REVIEW

File: gqcp/include/QCMethod/CC/CCD.hpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ class CCD {
         // Furthermore, the solvers only find the ground state wave function parameters, so the QCStructure only needs to contain the parameters for one state.
         const auto& T2 = environment.t2_amplitudes.back();
 
-        const auto E_electronic_correlation = environment.electronic_energies.back();
+        const auto E_electronic_correlation = environment.correlation_energies.back();
```
Comment (dariatols) [inline gqcp/include/QCMethod/CC/CCD.hpp:63]: This name is much clearer and reflects the purpose of this variable in a better way.

File: gqcp/include/QCMethod/CC/T2DIIS.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     *  Calculate the accelerated T2 amplitudes and place them in the environment by overwriting the previous T2 amplitudes.
+     * 
+     *  @param environment              The environment that acts as a sort of calculation space.
+     */
+    void execute(Environment& environment) override {
+
+        // Don't do anything if the minimum number of T2 amplitude iterations isn't satisfied.
+        if (environment.t2_amplitude_errors.size() < this->minimum_subspace_dimension) {
+            return;
```
Comment (dariatols) [inline gqcp/include/QCMethod/CC/T2DIIS.hpp:96]: Why don't we return an error if `t2_amplitude_errors.size() <minimum_subspace_dimension` happens?

File: gqcp/include/QCMethod/CC/T2ErrorCalculation.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     *  Calculate the current T2 error vector and add it to the environment.
+     * 
+     *  @param environment              The environment that acts as a sort of calculation space.
+     */
+    void execute(Environment& environment) override {
+
+        // Read the last two T2 amplitudes iterations and calculate the error as their difference.
+        const auto second_to_last_it = environment.t2_amplitudes.end() - 2;
```
Comment (dariatols) [inline gqcp/include/QCMethod/CC/T2ErrorCalculation.hpp:66]: If we only want the last two T2 amplitudes, shouldn't we use `environment.t2_amplitudes.end() - 1` for the second to last element?

**Review by dariatols — COMMENTED** [review]
Looks great!

**Review by xdvriend — APPROVED** [review]
👍

File: gqcp/include/QCMethod/CC/T2DIIS.hpp (hunk 2 above)
Comment (lelemmen) [reply gqcp/include/QCMethod/CC/T2DIIS.hpp:96]: We only want to _activate_ the DIIS when the minimum subspace dimension has been reached.

File: gqcp/include/QCMethod/CC/T2ErrorCalculation.hpp (hunk 3 above)
Comment (lelemmen) [reply gqcp/include/QCMethod/CC/T2ErrorCalculation.hpp:66]: No, `.end()` returns an iterator that points to the element _after_ the end of a `vector` (or a `deque`). See also the included example.

Schermafbeelding 2021-02-19 om 14 39 52

Comment (dariatols) [reply gqcp/include/QCMethod/CC/T2ErrorCalculation.hpp:66]: @lelemmen I was not aware of this, good to know, thanks!

**Pull request merged**